src/server/pps/server/master.go

```diff
@@ -158,6 +166,31 @@ eventLoop:
 	}
 }
 
+func (m *ppsMaster) attemptStep(ctx context.Context, e *pipelineEvent) error {
+	var errCount int
+	return backoff.RetryNotify(func() error {
+		// Create/Modify/Delete pipeline resources as needed per new state
+		return m.step(e.pipeline, e.etcdVer, e.etcdRev)
+	}, backoff.NewInfiniteBackOff(), func(err error, d time.Duration) error {
+		var failError failPipelineError
+		if errors.As(err, &failError) {
+			// don't just return this error, we might have to fail the pipeline
+			if errCount++; errCount >= maxErrCount || !failError.retry {
+				failError := m.a.setPipelineFailure(ctx, e.pipeline, fmt.Sprintf(
+					"could not update resources after %d attempts: %v", errCount, err))
+				if failError != nil {
+					return fmt.Errorf("error failing pipeline %q: %v", e.pipeline, failError)
+				}
+				return fmt.Errorf("failing pipeline %q: %v", e.pipeline, err)
```

Comment:
Update: after reading the `PIPELINE_FAILURE` section below, another idea I had was to change `failPipelineError` so that failure is optional and then rename it. Perhaps something like:
```
type errRetryStep struct {
  error  // go "inheritance" - field 'error' of type 'error'
  maxRetries int
  failPipelineOnExit bool
}
```
Then if `deletePipelineResources` errors, we can still retry the `step()` call by returning `errRetryStep`, but set `failPipelineOnExit` to `false`. Incidentally, this struct would remove the possible name conflict here.

Reply:
Agreed that the shadowing is unneeded, not sure why I went that direction. Adding the extra field does seem appropriate for capturing all of the information needed at the top level; I don't think my hesitance to add more was based on any good reason.

Reply:
I've left `retry` as a bool, I'm not entirely sure how the retry count logic would propagate if subsequent errors were different.